components/search/search.tsx

```diff
@@ -0,0 +1,392 @@
+// Copyright (c) 2015-present Mattermost, Inc. All Rights Reserved.
+// See LICENSE.txt for license information.
+
+import React, {ChangeEvent, FormEvent, useEffect, useState} from 'react';
+import {FormattedMessage} from 'react-intl';
+import classNames from 'classnames';
+
+import {searchHintOptions, RHSStates} from 'utils/constants';
+import * as Utils from 'utils/utils.jsx';
+import HeaderIconWrapper from 'components/channel_header/components/header_icon_wrapper';
+import SearchHint from 'components/search_hint/search_hint';
+import FlagIcon from 'components/widgets/icons/flag_icon';
+import MentionsIcon from 'components/widgets/icons/mentions_icon';
+import Popover from 'components/widgets/popover';
+import UserGuideDropdown from 'components/channel_header/components/user_guide_dropdown';
+
+import SearchBar from 'components/search_bar/search_bar';
+import SearchResults from 'components/search_results';
+import SearchDateProvider from 'components/suggestion/search_date_provider';
+import SearchChannelProvider from 'components/suggestion/search_channel_provider';
+import SearchUserProvider from 'components/suggestion/search_user_provider';
+
+import type {Props} from './types';
+
+interface SearchHintOption {
+    searchTerm: string,
+    message: {
+        id: string,
+        defaultMessage: string,
+    },
+}
+
+const determineVisibleSearchHintOptions = (searchTerms: string): SearchHintOption[] => {
+    let newVisibleSearchHintOptions: SearchHintOption[] = [];
+
+    if (searchTerms.trim() === '') {
+        return searchHintOptions;
+    }
+
+    const pretextArray = searchTerms.split(/\s+/g);
+    const pretext = pretextArray[pretextArray.length - 1];
+    const penultimatePretext = pretextArray[pretextArray.length - 2];
+
+    const shouldShowHintOptions = penultimatePretext ?
+        !searchHintOptions.some(({searchTerm}) => penultimatePretext.toLowerCase().endsWith(searchTerm.toLowerCase())) :
+        !searchHintOptions.some(({searchTerm}) => searchTerms.toLowerCase().endsWith(searchTerm.toLowerCase()));
+
+    if (shouldShowHintOptions) {
+        try {
+            newVisibleSearchHintOptions = searchHintOptions.filter((option) => {
+                return new RegExp(pretext, 'ig').
+                    test(option.searchTerm) && option.searchTerm.toLowerCase() !== pretext.toLowerCase();
+            });
+        } catch {
+            newVisibleSearchHintOptions = [];
+        }
+    }
+
+    return newVisibleSearchHintOptions;
+};
+
+const Search: React.FC<Props> = (props: Props): JSX.Element => {
+    const {actions, searchTerms} = props;
+
+    // generate intial component state and setters
+    const [focussed, setFocussed] = useState<boolean>(false);
+    const [keepInputFocussed, setKeepInputFocussed] = useState<boolean>(false);
+    const [indexChangedViaKeyPress, setIndexChangedViaKeyPress] = useState<boolean>(false);
```

Comment:
It is not ultimately affecting rendering, but since it does not trigger unnecessary re-renders it could stay like this.
We could otherwise utilise `useRef`, but it is less elegant